File: frontend/src/api/validazione.ts
/**
 * API client per la validazione di coerenza degli eventi di una partita.
 *
 * Diverso dalla ricostruzione: NON applica gli eventi al motore (non
 * crea snapshot), si limita a rilevare problemi di coerenza in un
 * colpo solo. Pensato per essere chiamato prima di "Ricostruisci"
 * per dare all'utente una checklist di problemi da correggere.
 */

import { chiamaApi, cliente } from "./cliente";

export type SeveritaProblema = "errore" | "avviso";

export type CodiceProblema =
  | "attacco_territori_non_adiacenti"
  | "attacco_da_territorio_non_posseduto"
  | "attacco_su_territorio_proprio"
  | "attacco_difensore_inesistente"
  | "doppio_turno_iniziato"
  | "evento_fuori_ordine_temporale"
  | "giocatore_id_inesistente"
  | "territorio_inesistente"
  | "armate_piazzate_su_territorio_altrui"
  | "spostamento_territori_non_adiacenti";

export interface ProblemaCoerenza {
  severita: SeveritaProblema;
  codice: CodiceProblema;
  messaggio: string;
  evento_id: string | null;
  posizione: number | null;
}

export interface RisultatoValidazioneCoerenza {
  n_eventi_analizzati: number;
  n_errori: number;
  n_avvisi: number;
  problemi: ProblemaCoerenza[];
}

export const apiValidazione = {
  /** Verifica problemi di coerenza degli eventi validati. */
  valida: (partitaId: string): Promise<RisultatoValidazioneCoerenza> =>
    chiamaApi(
      cliente.get<RisultatoValidazioneCoerenza>(
        `/partite/${partitaId}/valida-coerenza`,
      ),
    ),
};

File: backend/app/schemi/validazione.py
"""
Schemi per la validazione di coerenza degli `EventoValidato` di una
partita, **prima** della ricostruzione tramite motore.

Il motore (`risiko_engine`) solleva eccezioni quando uno stato è
incoerente, ma queste sono basate sull'ordine di applicazione e si
fermano al primo errore. Il validatore di coerenza scorre tutti gli
eventi e raccoglie **tutti** i problemi in un colpo solo, organizzati
per severità, così l'utente li vede tutti nella UI di review prima
di lanciare la ricostruzione.

Severità:
- `errore`: il motore certamente fallirà su questo evento (es. attacco
  fra territori non adiacenti, attacco da territorio non posseduto).
- `avviso`: comportamento sospetto ma applicabile (es. due TURNO_INIZIATO
  consecutivi senza eventi in mezzo, attacchi senza dadi difensori).
"""

from typing import Literal

from pydantic import BaseModel, Field

SeveritaProblema = Literal["errore", "avviso"]


CodiceProblema = Literal[
    "attacco_territori_non_adiacenti",
    "attacco_da_territorio_non_posseduto",
    "attacco_su_territorio_proprio",
    "attacco_difensore_inesistente",
    "doppio_turno_iniziato",
    "evento_fuori_ordine_temporale",
    "giocatore_id_inesistente",
    "territorio_inesistente",
    "armate_piazzate_su_territorio_altrui",
    "spostamento_territori_non_adiacenti",
]


class ProblemaCoerenza(BaseModel):
    """Singolo problema rilevato dal validatore."""

    severita: SeveritaProblema
    codice: CodiceProblema = Field(
        ..., description="Codice machine-readable del tipo di problema"
    )
    messaggio: str = Field(
        ..., description="Descrizione human-readable in italiano"
    )
    evento_id: str | None = Field(
        default=None,
        description="ID dell'EventoValidato problematico, None se globale",
    )
    posizione: int | None = Field(
        default=None,
        description="Posizione cronologica (0-indexed) dell'evento",
    )


class RisultatoValidazioneCoerenza(BaseModel):
    """Esito completo del validatore."""

    n_eventi_analizzati: int
    n_errori: int
    n_avvisi: int
    problemi: list[ProblemaCoerenza]

    @property
    def is_coerente(self) -> bool:
        """True se non ci sono errori (gli avvisi non bloccano)."""
        return self.n_errori == 0
